src/gt4py/storage/utils.py

```diff
@@ -105,54 +124,41 @@ def strides_from_padded_shape(padded_size, order_idx, itemsize):
     return list(strides)
 
 
-def allocate(default_origin, shape, layout_map, dtype, alignment_bytes, allocate_f):
-    dtype = np.dtype(dtype)
-    assert (
-        alignment_bytes % dtype.itemsize
-    ) == 0, "Alignment must be a multiple of byte-width of dtype."
-    itemsize = dtype.itemsize
-    items_per_alignment = int(alignment_bytes / itemsize)
-
-    order_idx = idx_from_order([i for i in layout_map if i is not None])
-    padded_shape = compute_padded_shape(shape, items_per_alignment, order_idx)
-
-    strides = strides_from_padded_shape(padded_shape, order_idx, itemsize)
-    if len(order_idx) > 0:
-        halo_offset = (
-            int(math.ceil(default_origin[order_idx[-1]] / items_per_alignment))
-            * items_per_alignment
-            - default_origin[order_idx[-1]]
-        )
+def get_ptr(array):
+    if isinstance(array, np.ndarray):
+        return array.ctypes.data
     else:
-        halo_offset = 0
-
-    padded_size = int(np.prod(padded_shape))
-    buffer_size = padded_size + items_per_alignment - 1
-    array, raw_buffer = allocate_f(buffer_size, dtype=dtype)
-
-    allocation_mismatch = int((array.ctypes.data % alignment_bytes) / itemsize)
+        return array.data.ptr
 
-    alignment_offset = (halo_offset - allocation_mismatch) % items_per_alignment
 
-    field = np.reshape(array[alignment_offset : alignment_offset + padded_size], padded_shape)
-    if field.ndim > 0:
-        field.strides = strides
-        field = field[tuple(slice(0, s, None) for s in shape)]
-    return raw_buffer, field
+def _reshape(array, strides, padded_shape):
+    if isinstance(array, np.ndarray):
+        field = np.reshape(array, padded_shape)
+        if field.ndim > 0:
+            field.strides = strides
+    else:
+        field = cp.reshape(array, padded_shape)
+        if field.ndim > 0:
+            field = as_strided(field, strides=strides)
+    return field
 
 
-def allocate_gpu_unmanaged(default_origin, shape, layout_map, dtype, alignment_bytes):
+def allocate(default_origin, shape, layout_map, dtype, alignment_bytes, allocate_f):
```

Comment:
As a general comment, it is a little dangerous to have so many positional arguments to a function. When these get refactored, it is easy to introduce bugs where arguments are shifted off by one position. This can be avoided by using keyword arguments, or required keyword arguments (e.g. `allocate(*, default_origin, shape, ...)`).